### PR TITLE
adding workspace_build resource

### LINKED ISSUE #4625: audit: log workspace start and stop actions
When a user starts or stops a workspace, either via the UI or CLI, we should generate audit logs.

### PR DESCRIPTION
**Pull request by Kira-Pilot**
resolves #4625 

We now log start/stop for workspace builds
Screen Shot 2022-10-21 at 2 33 16 PM

### CODE REVIEW

File: coderd/audit/request.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ func ResourceTarget[T Auditable](tgt T) string {
 		return typed.Username
 	case database.Workspace:
 		return typed.Name
+	case database.WorkspaceBuild:
+		return string(typed.Transition)
```
Comment (Kira-Pilot) [inline coderd/audit/request.go:47]: This need to be the workspace name

File: coderd/audit/request.go [hunk 2]
```diff
@@ -81,6 +85,8 @@ func ResourceType[T Auditable](tgt T) database.ResourceType {
 		return database.ResourceTypeUser
 	case database.Workspace:
 		return database.ResourceTypeWorkspace
+	case database.WorkspaceBuild:
+		return database.ResourceTypeWorkspaceBuild
```
Comment (Kira-Pilot) [inline coderd/audit/request.go:89]: This is missing - must generate in `dump.sql`

**Comment by Kira-Pilot** [comment]
@coadler I'm close with this!
Screen Shot 2022-10-19 at 1 15 10 PM

But I want to get the workspace name in this friendly string, which does not exist on the `workspace_build` resource.

I figured I would just add an optional argument to `InitRequest`: `AlternativeResource` and pass that through; however, it seems there aren't optional arguments in Go. Also, maybe that's ugly.

Do you have any ideas?

**Comment by coadler** [comment]
Audit logs have a field called `additional_resources`, which is intended for this scenario. I haven't really piped it through yet though. This is where you could put the workspace name!

File: coderd/audit/request.go [hunk 3]
```diff
@@ -43,6 +44,9 @@ func ResourceTarget[T Auditable](tgt T) string {
 		return typed.Username
 	case database.Workspace:
 		return typed.Name
+	case database.WorkspaceBuild:
+		// this isn't used
+		return string(typed.BuildNumber)
```
Comment (Kira-Pilot) [inline coderd/audit/request.go:49]: Ug tell me there's a better way.

File: coderd/audit.go [hunk 4]
```diff
@@ -220,12 +220,26 @@ func convertAuditLog(dblog database.GetAuditLogsOffsetRow) codersdk.AuditLog {
...
 	)
 
+	// Strings for build updates follow the below format:
+	// "{user} started workspace build for workspace {target}"
+	// where target is a workspace instead of the workspace build
+	if alog.ResourceType == database.ResourceTypeWorkspaceBuild {
+		workspace_bytes := []byte(alog.AdditionalFields)
+		var workspaceResourceInfo WorkspaceResourceInfo
+		json.Unmarshal(workspace_bytes, &workspaceResourceInfo)
+		str += " for workspace " + workspaceResourceInfo.WorkspaceName
```
Comment (Kira-Pilot) [inline coderd/audit.go:240]: I would prefer not to add this here and instead to use the `ResourceTarget` function.

**Review by BrunoQuaresma — APPROVED** [review]
FE and BE code looks good to me! I don't have answers for your questions tho.

File: coderd/database/migrations/000065_add_resource_type_workspace_build.up.sql [hunk 5]
```diff
@@ -0,0 +1 @@
+ALTER TYPE resource_type ADD VALUE IF NOT EXISTS 'workspace_build';
```
Comment (coadler) [inline coderd/database/migrations/000065_add_resource_type_workspace_build.up.sql:1]: Could we just collapse these into a single migration?

File: coderd/templates.go [hunk 6]
```diff
@@ -84,10 +85,11 @@ func (api *API) deleteTemplate(rw http.ResponseWriter, r *http.Request) {
...
 		aReq, commitAudit = audit.InitRequest[database.Template](rw, &audit.RequestParams{
-			Audit:   auditor,
-			Log:     api.Logger,
-			Request: r,
-			Action:  database.AuditActionDelete,
+			Audit:            auditor,
+			Log:              api.Logger,
+			Request:          r,
+			Action:           database.AuditActionDelete,
+			AdditionalFields: json.RawMessage("{}"),
```
Comment (coadler) [inline coderd/templates.go:92]: I think I would prefer if we checked if this was nil in the `audit.InitRequest` so we didn't have to provide this default in each handler!

File: coderd/workspacebuilds.go [hunk 7]
```diff
@@ -278,28 +278,64 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
...
+
+		var (
+			aReq, commitAudit = audit.InitRequest[database.WorkspaceBuild](rw, &audit.RequestParams{
+				Audit:            *auditor,
+				Log:              api.Logger,
+				Request:          r,
+				Action:           auditAction,
+				AdditionalFields: wriBytes,
+			})
+		)
```
Comment (coadler) [inline coderd/workspacebuilds.go:328]: small nit: since this is just a single variable you don't really need the `var ()` block, you can just use `:=`.

**Review by coadler — APPROVED** [review]

**Pull request merged**